ext/test/fail_fs/fail_fs.c

```diff
@@ -239,10 +239,12 @@ fail_file_read(
     } else
         read_ops = ++fail_fs->read_ops;
 
-    fail_fs_unlock(&fail_fs->lock);
-
-    if (fail_fs->fail_enabled && fail_fs->allow_reads != 0 && read_ops % fail_fs->allow_reads == 0)
+    if (fail_fs->fail_enabled && fail_fs->allow_reads != 0 &&
+      read_ops % fail_fs->allow_reads == 0) {
+        fail_fs_unlock(&fail_fs->lock);
```

Comment:
As discussed off line for both this function and "write", I'd suggest capturing the return code to `ret` and jumping (goto) to a common return path at the end of the function.